### PR TITLE
Add statevector support and basic error checking to plot_state

### LINKED ISSUE #907: Make plot_state automatically convert state vectors to density matrices
### What is the expected enhancement?

When calling the `plot_state` visualization function the input must be a density matrix, however the statevector simulator returns a state vector that must  be manually converted to a density matrix. This should be done automatically when a user calls `plot_state(statevector)`:

Expected use:

```python
import qiskit
from qiskit.tools.visualization import plot_state
qr = qiskit.QuantumRegister(2)
circ = qiskit.QuantumCircuit(qr)
circ.h(qr[0])
circ.cx(qr[0], qr[1])
result = qiskit.execute(circ, 'local_statevector_simulator).result()
state = result.get_statevector(circ)
fig = plot_state(state)
fig.show()
```

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Fixes #907 by automatically converting a statevector to a density matrix when calling `plot_state`.

### Details and comments

This PR also adds basic error handling to check that the input (after conversion to a density matrix) is a square matrix, and is an n-qubit state.

### CODE REVIEW

File: qiskit/tools/visualization/_state_visualization.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 from matplotlib.patches import FancyArrowPatch
 from mpl_toolkits.mplot3d import proj3d
 import numpy as np
+from qiskit import QISKitError
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: We did recently add a VisualizationError class, I'm wondering if that would be more useful here.

https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/tools/visualization/_error.py#L13

What do you think?

Comment (chriseclectic) [reply qiskit/tools/visualization/_state_visualization.py]: I didn't know about that one, I'll change it over.

**Comment by mtreinish** [comment]
One other random thought I just had is maybe we should move/rename the qiskit.tools.qi.qi.outer helper function to somewhere more obvious because we do this operation in a lot of places. It's only a oneline operation but considering it's called in a bunch of places that might be a useful things in the future. But, not something worth blocking this on.

**Comment by chriseclectic** [comment]
@mtreinish I think we should replace`qiskit.tools.qi.qi.outer(arg)` everywhere in the source code with `np.outer(arg, np.conj(arg))` since it's shorter and doesn't need an additional import from some deeply nested module.

**Pull request merged**